step/src/main/java/com/google/sps/DatabaseHandler.java

```diff
@@ -15,43 +15,135 @@
 package com.google.sps;
 
 import java.util.Collection;
-import java.util.Map;
-import java.util.HashMap;
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.NoSuchElementException;
+import javax.servlet.http.HttpServletRequest;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.SortDirection;
+import com.google.appengine.api.datastore.Query.CompositeFilter;
+import com.google.appengine.api.datastore.Query.CompositeFilterOperator;
+import com.google.appengine.api.datastore.Query.Filter;
+import com.google.appengine.api.datastore.Query.FilterOperator;
+import com.google.appengine.api.datastore.Query.FilterPredicate;
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.Key;
+import com.google.appengine.api.datastore.KeyFactory;
+import com.google.appengine.api.datastore.EntityNotFoundException;
 
 public final class DatabaseHandler {
   
-  // TODO: replace with Datastore
-  private static Map<Long, User> idMap = new HashMap<>();
+  // Accessing Datastore
+  private static DatastoreService datastore =
+         DatastoreServiceFactory.getDatastoreService();
 
   private DatabaseHandler() {
 
   }
 
-  // TODO: call this method when a user first signs up
   // Method for adding a user to the database
-  public static void addUserToDatabase(User user) {
-    idMap.put(user.getId(), user);    
+  public static void addUser(String firstName, String lastName, int dayBirth,
+    int monthBirth, int yearBirth, String email, String id) {
+    Entity entity = new Entity("User", id);
+    entity.setProperty("firstName", firstName);
+    entity.setProperty("lastName", lastName);
+    entity.setProperty("dayBirth", dayBirth);
+    entity.setProperty("monthBirth", monthBirth);
+    entity.setProperty("yearBirth", yearBirth);
+    entity.setProperty("email", email);
+    entity.setProperty("id", id);
+    datastore.put(entity);    
   }
+ 
+  // Adding a user's notificaton to the databse
+  public static void addNotification(String firstId, String secondId,
+      long timestamp, String type) {
```

Comment:
Can you make variables for type into a static constant? Something like DatabaseHandler.MATCHING being passed as a parameter here will significantly help prevent invalid input that could mess up the data base. Maybe an enum would work too. Also, since we're on the topic, validating the type parameter would be a good idea too.

Reply:
That makes sense! When you say validating the type parameter, do you mean checking that the type int that's passed in is one of the valid options?

Reply:
Yeah, exactly. That way the code doesn't run on parameter values that could cause problems.